cql3/untyped_result_set.hh

```diff
@@ -188,8 +217,19 @@ public:
         return _rows.back();
     }
 private:
+    using qr_tuple = std::tuple<foreign_ptr<lw_shared_ptr<query::result>>, shared_ptr<const cql3::metadata>>;
+    using storage = std::variant<std::monostate
+        , ::shared_ptr<cql_transport::messages::result_message>
+        , qr_tuple
+    >;
+    struct visitor;
```

Comment:
Why do we have to expose this visitor in the header? Wouldn't it be enough to make it internal to .cc file?

Reply:
namespacing + private types + friend relationship.

Reply:
I don't follow. Could you elaborate a bit more for weaker audience ;).

Reply:
The visitor/static below use types and constructors that I made private in the row type (because they should be internal). untyped_result_set can see them because it is a friend. Top-level routines/types cannot. Can make more elaborate friend declarations, but to me at least it seems easier to just do these harmless (and private) declarations instead. 
Also, I prefer nested types to file-local ones, if for nothing else so for scoping of concern.

Reply:
Got you. I was looking at a wrong friendship relation (result_generator <-> untyped_result_set) and couldn't figure out why it matters :)

Reply:
I agree about nesting, but the #include dependency addition is unfortunate. Maybe hide the variant behind a unique_ptr? It's just a placeholder after all, no?

Also, std::any? probably doesn't work due to copyablity.

Reply:
I felt that yet another level of indirection was perhaps pushing it... Not happy about the extra indirection - really with the foreign_ptr was not so an lw_shared_ptr... Even tried to work around the forward declaration aspect there, but it just becomes terrible. 
std::any does not work because of move-only types.